p2p: never check tx rejections by txid

glozow opened the pull request:
This PR removes the rejection cache filtering part of all txid-based tx requests. In practice, this seems to only be used for orphan resolution because everybody supports wtxidrelay. And so in essence, this PR removes the logic that stops us from attempting orphan resolution when parents are found in the rejection filter.

Background:
We have 2 bloom filters for remembering transactions that we've rejected so we don't redownload them, `RecentRejectsFilter` and `RecentRejectsReconsiderableFilter`. This lets us save a little bit of bandwidth on rejected transactions, particularly if we have policy differences. It's not designed to stop attackers from wasting our download bandwidth, as they can create as many policy-invalid transactions with different witnesses as they want. We generally only put wtxids and not txids in them (see #18044), except for these cases:
(A) When we specifically know that the rejection reason is not due to the witness (however, this is only done for `!AreInputsStandard`, even though there are other cases we could apply)
(B) When the transaction has no witness, so its txid == wtxid
(C) (With #32379) When the transaction's witness has been stripped, so txid == wtxid
We check the filters to decide whether to send a getdata for an inv, whether we should validate a tx we just downloaded, and whether we should keep an orphan: we look the missing parents up by (prevout) txid and throw the orphan away if its parents were already rejected. That means there are very few cases where we save bandwidth by finding a txid in the filter.

Rationale: mainly that the additional complexity is not worth the bandwidth savings. TLDR: this case does not seem to ever hit in practice. Even if that is circumstantial, the number of transactions this filter could apply to is extremely small.

I collected some stats over a couple of weeks, setting mempoolexpiry to 1 hour to try to artificially increase orphan rates (though I don't know whether the effect is significant):
- I'm seeing 100% of peers support wtxidrelay. Bitcoin Core has been doing this since ~5 years ago. I'm sure there are some nodes that don't do it, but it'll probably be rare that we are surrounded by txidrelay peers and that we have a policy difference causing us to reject the transaction that they all send us.
- Orphan parent fetching was 3.64% of my transaction requests.
- 93.77% of all transactions I receive have witnesses (includes accepted and rejected ones). 96.20% of orphan parents I fetched had witnesses.
- I found no cases of already-rejected parent txids in the past week. This could be because all my peers have the same policy as me (which might not always be true), but the number of nonsegwit orphan parents in general is just so tiny - a few dozen requests per day.

glozow committed:
[p2p] never check rejections by txid in AlreadyHaveTx

While this may result in some wasted bandwidth for orphans that have
invalid non-witness parents, we expect this case to be rare. Allows us
to cache witness-stripped transactions in recent rejections without
worrying about blocking 1p1cs.

glozow commented:
cc @darosior

darosior commented:
Concept ACK

cedwies reviewed (commented):
Built PR 33066 on macOS 14 (Apple Silicon) with CMake + Ninja, Debug.
- unit tests: 141/141 pass, 9 min wall time

what I checked:
‒ AlreadyHaveTx(): tx-id path now skips both reject filters; wtx-id path unchanged.  
‒ MempoolRejectedTx(): parent-txid rejection bail-out removed → orphans kept even if parents sit in a reject cache 
‒ updated tests (expected_behaviors table + orphan-handling cases) run green.

Question (still wrapping my head around this):
MempoolRejectedTx still inserts a tx-id into RecentRejectsFilter for TX_INPUTS_NOT_STANDARD, but After this patch AlreadyHaveTx no longer consults the filter by tx-id. Is that entry now redundant, or do other call-sites still depend on it?

(Second review, so no (N)ACK yet. just confirming I understand the (PR) flow.)

darosior commented:
> I didn't realize it was always meant to go - how do you know that?

It was introduced in https://github.com/bitcoin/bitcoin/pull/18044 as a hack until the network upgrades to wtxid relay with an explicit mention it can be (according to author, [should be](https://github.com/bitcoin/bitcoin/pull/18044#discussion_r388445400)) removed afterward:
https://github.com/bitcoin/bitcoin/blob/8a94cf8efebc3177effcfc1160560735b8caf34b/src/node/txdownloadman_impl.cpp#L452-L454

Of course that it was *meant to* go then does not implies it *needs to* go now. But i think this still holds. It is a weird special case that is not necessary post wtxid relay (and post this PR).

darosior commented:
I opened https://github.com/bitcoin/bitcoin/pull/33105, which ended up implementing a variant of @ajtowns' suggestion, because as pointed out the resource consumption gains can be achieved independently of this work (which i still think is desirable) as a much smaller patch which i think would be nice to get in before the feature freeze.

ajtowns commented:
> However it seemed cleaner to me to get rid of the WITNESS_STRIPPED edge case detection in the first place, which was always meant to go?

I don't think it makes sense to get rid of WITNESS_STRIPPED while we still care about resolving orphans via their missing parents' txids. If/when we have protocol support for receiver-initiated package relay by wtxid, then I think it could make sense to treat every request as being by wtxid -- so a witness-stripped response from a non-wtxid-relay peer would mean not requesting that tx from any other non-wtxid-relay peer, but wouldn't prevent requesting the same tx from wtxid-relay peers.

Crypt-iQ commented on src/node/txdownloadman_impl.cpp:
```diff
@@ -361,83 +366,47 @@ node::RejectedTxTodo TxDownloadManagerImpl::MempoolRejectedTx(const CTransaction
-                if (RecentRejectsFilter().contains(parent_txid.ToUint256())) {
-                    fRejectedParents = true;
-                    break;
-                } else if (RecentRejectsReconsiderableFilter().contains(parent_txid.ToUint256()) &&
```
I get that if this is an orphan, we only have the parent txids. In most cases, it seems like this reconsiderable filter will be storing wtxid? The exceptions being a non-segwit tx and if a parent is TX_RECONSIDERABLE but is actually witness-stripped (I'm not sure if there are others?). I may be missing something here, are there more cases where this reconsiderable filter check be used with txid?

glozow replied:
Yep, we don't ever put txids in the reconsiderable filter unless the txid is the same as the wtxid. We don't put witness-stripped in there either (even if we do #32379, as that will put it in the normal RecentRejects filter)

Crypt-iQ replied:
I modified p2p_opportunistic_1p1c.py to remove the witness on a low-fee parent tx and it failed with TX_RECONSIDERABLE (as those checks are before the 3x CheckInputScripts that returns WITNESS_STRIPPED). I think my comment was a little confusing, also not sure this matters in practice?

glozow commented:
I've removed #32379 and mention of witness-stripping from the PR description; I think we can regard this as irrelevant for removing the triple validation (though it doesn't hurt).

The primary motivation now is to clean up logic that doesn't really get used.

> Others include base size > 100k, non-standard output types, min feerate not met (when counting just base size), too many sigops, and probably a few more.

Also: large scriptSig, nonstandard output type, dust, OP_RETURN size and count (assuming this node's policy is more restrictive than its peers'). I would guess that if we added these, they would similarly be very rare and not really worth the added complexity.

We can wait until we're using something like BIP331 for orphan resolution (when txid requests are even rarer), but I think there's sufficient evidence this is already vestigal today.

ajtowns commented:
sipa wrote:

> * For **a specific, small subset** of policy-invalidity reasons that do not involve the witness (namely: nonstandard input types, because that's currently the only thing where we assign failure to the txid, not the wtxid) or are non-segwit.

I believe you'd see that for new segwit spend types when you're not running an upgraded node (eg, taproot, p2a, perhaps p2qrh), which might help reduce your wasted bandwidth if you're running behind network consensus/policy changes as adoption increases.

> As an alternative, it is possible to expand the number of validation failures that get attributed to txids, as there are more than just non-standard script input types in theory. Others include base size > 100k, non-standard output types, min feerate not met (when counting just base size), too many sigops, and probably a few more. I think those are still unlikely candidates of being things people build at scale, in dependent transactions, but if we think that's possible, I think expanding those to be txid-rejectable is a reasonable alternative to this PR.

Bumping tx version (TRUC), creating dust outputs (ephemeral dust), and increased OP_RETURN output count/size are all pretty recent examples of where we've loosened standardness rules in ways that could be assigned to the txid and that presumably we're hoping might get built on at scale. I don't see why we'd necessary stop doing that any time soon.

Of course, all those things will only show up in normal usage if your mempool rules are stricter than what's being commonly seen on the network, and stricter than your peers, which is a situation we generally try to avoid. So even then, I don't think you'll see much effect in normal usage -- it's only defensive against future changes to policy that you don't adopt in sync with your peers, or if there's something of a schism in the p2p network and we care about the people adopting more restrictive policies.

It might be interesting to try getting similar stats on a Knots node with its more restrictive mempool policies -- potentially both just seeing how many times rejected txs get redundantly downloaded due to orphan resolution attempts, and tweaking it so that (some of) the errors that can be assigned to the txid are (which could catch the lower datacarrier limits, though not the inscription ones)... Alternatively maybe you could get similar stats from a post-wtxidrelay but pre-taproot node who might see child spends of taproot parents.

That doesn't help for non-standardness issues we can't assign to the txid -- so script upgrades and forbidding inscriptions and they like aren't helped by this mechanism, and can't be helped without some form of explicit package relay afaics.

The bad case there is if we get a child from many peers whose parent we've already probably rejected (by wtxid). If we can also reject the txid, then we save ourselves some bandwidth and perhaps validation costs; if we can't reject the txid, we likely have to redundantly request it from each peer in case many of our peers are dishonest, but one is honest and has a version of the tx that we'll accept. So rather than downloading it once, we're downloading it up to 8 or 120 times if all our peers have looser standardness policies than we do? We already check the wtxid against RecentRejects via AlreadyHaveTx via ReceivedTx, so we won't revalidate afaics.

As far as the adversarial case goes -- the worst case would be that in an environment where your rules are stricter than many of your peers, an adversary could force you to redownload/revalidate a large tx from many peers with you rejecting it each time. If your peers announced the tx by wtxid, you'd only do that once, and things would be fine. To trigger it happening by txid, they would need to use orphan resolution and actually have each of your peers announce the child (in order for you to believe its worth getting the parent from them). But that's just normal behaviour -- you're sending a real child tx through the network?

So I think the only potential benefit here is saving ourselves from redownloading a transaction ~10x or ~100x times more than we should, and agree that this is only in the cases where we're enforcing "special" standardness rules, and those rules can be assigned independent of the witness data, and only in cases where we have a child of the non-standard tx.

> Absent that alternative, I feel the hypothetical scenario where this matters is just too obscure to have a whole piece of net_processing state dedicated too.

I don't really think this change is much of a win -- the original motivation of avoiding the triple-script-check stuff is, but that can be avoided just by rejecting witness-stripped txs asap, independent of this change.

>  It uses memory too; without inserting txids in the rejection filter, we could shrink the filters, or improve their reliability.

I don't think that's quite true in practice? The logic in txdownloadman_impl.h for the the filter size is based on 1000 txs per second for 2 minutes creating 120000 entries; but that should presumably be doubled if we're assuming an attacker might add minimal witness data to invalid non-standard input txs as well in order to overflow the filter as quickly as possible in order to trick us into re-requesting txs from other peers? If we're approximately never seeing spends of non-standard inputs, then we also already have the reliability benefits of never adding them. Either way, 1.3MB, 2.6MB or 650kB of filters doesn't seem like a big deal to me. If we really cared about minimising the filter size we could probably rate-limit peers to a target of 10 or 20 rejected txs per second or similar.

glozow wrote:

> We can wait until we're using something like BIP331 for orphan resolution (when txid requests are even rarer), but I think there's sufficient evidence this is already vestigal today.

I think the main reason it seems vestigial is dependent on running a node with policy rules at least as loose as (almost) all your peers, and assuming that everyone else who isn't doing the same now, will do so in short order.

ajtowns commented:
> and/or consider preferentially peering with wtxidrelay peers

It would probably be good to be able to preferentially peer with package-relay peers when they exist, so even if peering with wtxidrelay peers is easy now, having the code in a way that's reusable for package-relay peers in future might be worthwhile.
